### ISSUE TITLE
Test failures

### ISSUE THREAD

**Issue by Luflosi** [kovidgoyal/kitty#4175]
**Describe the bug**
On my MacBook with macOS 10.15 Catalina and in my macOS VM with 10.13 High Sierra `make debug && make test` fails when using the dependencies from a Nix shell but not on an iMac with macOS 11 Big Sur using Homebrew. On the MacBook when I exit the nix-shell, remove anything Nix related from the PATH, install everything via Homebrew and rebuild again, the test does not fail.
MacBook:
```
======================================================================
FAIL: test_file_put (kitty_tests.file_transmission.TestFileTransmission)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/path/to/kitty/kitty_tests/file_transmission.py", line 304, in test_file_put
    self.ae(st.st_mtime_ns, 13)
    c = <zlib.Compress object at 0x105cfa5b0>
    dest = '/private/var/folders/c5/yg8qdfhx76l30_7wkjm9h3gw0000gn/T/tmpglwnt9ai/2.bin'
    f = <_io.TextIOWrapper name='/private/var/folders/c5/yg8qdfhx76l30_7wkjm9h3gw0000gn/T/tmpi76u0a68/3' mode='r' encoding='UTF-8'>
    fid = 1
    ft = <kitty.file_transmission.TestFileTransmission object at 0x1052eafa0>
    one = '/private/var/folders/c5/yg8qdfhx76l30_7wkjm9h3gw0000gn/T/tmpi76u0a68/1'
    quiet = 2
    self = <kitty_tests.file_transmission.TestFileTransmission testMethod=test_file_put>
    send = <function TestFileTransmission.test_file_put.<locals>.send at 0x105af8f70>
    st = os.stat_result(st_mode=33279, st_ino=8739274087, st_dev=16777220, st_nlink=1, st_uid=501, st_gid=20, st_size=3, st_atime=0, st_mtime=0, st_ctime=1635673897)
    three = '/private/var/folders/c5/yg8qdfhx76l30_7wkjm9h3gw0000gn/T/tmpi76u0a68/3'
    two = '/private/var/folders/c5/yg8qdfhx76l30_7wkjm9h3gw0000gn/T/tmpi76u0a68/2'
AssertionError: 0 != 13
```
VM (two more failures):
```
======================================================================
ERROR: test_path_mapping_receive (kitty_tests.file_transmission.TestFileTransmission)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/path/to/kitty/kitty_tests/file_transmission.py", line 388, in test_path_mapping_receive
    tf([b/'r', b/'d'], {b/'r': b/'r', b/'d': b/'d', b/'d'/'r': b/'d'/'r'})
    am = <function TestFileTransmission.test_path_mapping_receive.<locals>.am at 0x10af655e0>
    b = PosixPath('/private/tmp/tmpx0rxbmsn/b')
    gm = <function TestFileTransmission.test_path_mapping_receive.<locals>.gm at 0x10af65670>
    opts = <kitty.cli_stub.CLIOptions object at 0x10b186ac0>
    self = <kitty_tests.file_transmission.TestFileTransmission testMethod=test_path_mapping_receive>
    tf = <function TestFileTransmission.test_path_mapping_receive.<locals>.tf at 0x10af65700>
  File "/path/to/kitty/kitty_tests/file_transmission.py", line 379, in tf
    files, specs = gm(all_specs)
    all_specs = [PosixPath('/private/tmp/tmpx0rxbmsn/b/r'), PosixPath('/private/tmp/tmpx0rxbmsn/b/d')]
    am = <function TestFileTransmission.test_path_mapping_receive.<locals>.am at 0x10af655e0>
    args = [PosixPath('/private/tmp/tmpx0rxbmsn/b/r'), PosixPath('/private/tmp/tmpx0rxbmsn/b/d')]
    dest = ''
    different_home = ''
    expected = {PosixPath('/private/tmp/tmpx0rxbmsn/b/r'): PosixPath('/private/tmp/tmpx0rxbmsn/b/r'), PosixPath('/private/tmp/tmpx0rxbmsn/b/d'): PosixPath('/private/tmp/tmpx0rxbmsn/b/d'), PosixPath('/private/tmp/tmpx0rxbmsn/b/d/r'): PosixPath('/private/tmp/tmpx0rxbmsn/b/d/r')}
    gm = <function TestFileTransmission.test_path_mapping_receive.<locals>.gm at 0x10af65670>
    opts = <kitty.cli_stub.CLIOptions object at 0x10b186ac0>
    self = <kitty_tests.file_transmission.TestFileTransmission testMethod=test_path_mapping_receive>
  File "/path/to/kitty/kitty_tests/file_transmission.py", line 368, in gm
    raise x
    all_specs = [PosixPath('/private/tmp/tmpx0rxbmsn/b/r'), PosixPath('/private/tmp/tmpx0rxbmsn/b/d')]
    files = []
    specs = [('0', '/private/tmp/tmpx0rxbmsn/b/r'), ('1', '/private/tmp/tmpx0rxbmsn/b/d')]
    x = TransmissionError('No permission to read spec')
kitty.file_transmission.TransmissionError: No permission to read spec

======================================================================
FAIL: test_file_get (kitty_tests.file_transmission.TestFileTransmission)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/path/to/kitty/kitty_tests/file_transmission.py", line 170, in test_file_get
    self.ae(len(files), 6)
    cwd = '/private/tmp/tmpwlv8inza/cwd'
    f = <_io.TextIOWrapper name='/private/tmp/tmpwlv8inza/home/a' mode='w' encoding='UTF-8'>
    f2 = <_io.TextIOWrapper name='/private/tmp/tmpwlv8inza/home/ad/b' mode='w' encoding='UTF-8'>
    files = {}
    ft = <kitty.file_transmission.TestFileTransmission object at 0x10b186370>
    home = '/private/tmp/tmpwlv8inza/home'
    quiet = 2
    self = <kitty_tests.file_transmission.TestFileTransmission testMethod=test_file_get>
AssertionError: 0 != 6

======================================================================
FAIL: test_file_put (kitty_tests.file_transmission.TestFileTransmission)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/path/to/kitty/kitty_tests/file_transmission.py", line 304, in test_file_put
    self.ae(st.st_mtime_ns, 13)
    c = <zlib.Compress object at 0x10b157d40>
    dest = '/private/tmp/tmpj8ojb8_f/2.bin'
    f = <_io.TextIOWrapper name='/private/tmp/tmph2_ijrq3/3' mode='r' encoding='UTF-8'>
    fid = 1
    ft = <kitty.file_transmission.TestFileTransmission object at 0x10b1866d0>
    one = '/private/tmp/tmph2_ijrq3/1'
    quiet = 2
    self = <kitty_tests.file_transmission.TestFileTransmission testMethod=test_file_put>
    send = <function TestFileTransmission.test_file_put.<locals>.send at 0x10af655e0>
    st = os.stat_result(st_mode=33279, st_ino=17519513, st_dev=16777223, st_nlink=1, st_uid=501, st_gid=0, st_size=3, st_atime=0, st_mtime=0, st_ctime=1635673826)
    three = '/private/tmp/tmph2_ijrq3/3'
    two = '/private/tmp/tmph2_ijrq3/2'
AssertionError: 0 != 13
```

**To Reproduce**
Fails reproducibly on the first two machines with `nix-shell --pure` and then `make debug && make test`.

**Environment details**
Using the latest kitty master at the time of writing (22673ebd90b1654f7c396485fa3bd4294967eed6).

**Additional context**
If you like, I can give you SSH access to the VM so you don't need to install Nix yourself.

**Comment by kovidgoyal** [comment]
What on earth does nix do? That indicates the utime() system call is failing.
Notice how the ctime is set in the stat result but both atime and mtime
are not. Does nix mount the filesystem with some options to prevent
mtime being recorded?

**Comment by Luflosi** [comment]
The only mount point related to Nix on my machine is /nix and I'm building and testing kitty with the normal `make debug && make test`, so I don't see how Nix could be changing any mounts during the build.
I tried building outside of a Nix shell with Python still installed globally with Nix and the build failure persists. If I remove the version of Python installed by Nix, the test no longer fails. Installing it again brings back the test failure. So this tells me that Nix modifies Python in an undesired way for some reason. I'll say this is a Nix problem for now and will try to find out more.
Do you know what the other two test failures in the VM are?

**Comment by kovidgoyal** [comment]
On Sun, Oct 31, 2021 at 03:56:29AM -0700, Luflosi wrote:
> The only mount point related to Nix on my machine is /nix and I'm building and testing kitty with the normal `make debug && make test`, so I don't see how Nix could be changing any mounts during the build.
> I tried building outside of a Nix shell with Python still installed globally with Nix and the build failure persists. If I remove the version of Python installed by Nix, the test no longer fails. Installing it again brings back the test failure. So this tells me that Nix modifies Python in an undesired way for some reason. I'll say this is a Nix problem for now and will try to find out more.

Possibly it breaks os.utime() in python. Try running some simple code
that tests os.utime() with that python to check.

> Do you know what the other two test failures in the VM are?

No the trace there doesnt tell  us what the underlying failure is, one
is a permission denied, which coms from line 100 in file_transmission.py
you can put some prints in there to get more info.

The other one could be anything, its saying no file metadata was sent, 
the file metadata is read in iter_file_metadata() put some prints there
to see whats going on.

**Comment by Luflosi** [comment]
Using
```py
import os
a = '/path/to/some/file'
os.utime(a, ns=(12, 34))
os.stat(a).st_mtime_ns
```
I determined that `os.utime()` from python installed via Nix seems to round the time down to the nearest thousand nanoseconds while python from Homebrew does not. I created https://github.com/NixOS/nixpkgs/issues/143987.
I'll try to debug the two other failures some other time.

**Comment by kovidgoyal** [comment]
Closing since I dont think this is a bug in the tests or the tested code. But if you have any new finsings feel free to update.

**Comment by Luflosi** [comment]
Apparently this is caused by Nix still using the macOS 10.12 SDK, which does not support a certain function required for nanosecond resolution, in which case Python falls back to microsecond resolution. For more details see the Nixpkgs issue linked above. I see you already updated the test to use a multiple of 1000 ns. I think this is a good solution as I think kitty still supports macOS 10.12 and there might be other systems where nanosecond resolution is not available. I tested it again and now this particular part of the test runs but there are a couple more instances where the test relies on nanosecond resolution. I submitted a PR to fix this.
